src/library_webgpu.js

```diff
@@ -2673,13 +2688,90 @@ var LibraryWebGPU = {
     sampler.label = UTF8ToString(labelPtr);
   },
 
+  // WGPUSurfaceCapabilities
+
+  wgpuSurfaceCapabilitiesFreeMembers: (value) => {
+    // wgpuSurfaceCapabilities does currently allocate anything
+  },
+
   // WGPUSurface
 
+  wgpuSurfaceConfigure: (surfaceId, config) => {
+    {{{ gpu.makeCheckDescriptor('config') }}}
+    var deviceId = {{{ makeGetValue('config', C_STRUCTS.WGPUSurfaceConfiguration.device, '*') }}};
+    var context = WebGPU.mgrSurface.get(surfaceId);
+
+#if ASSERTIONS
+    assert({{{ gpu.PresentMode.Fifo }}} ===
+      {{{ gpu.makeGetU32('config', C_STRUCTS.WGPUSurfaceConfiguration.presentMode) }}});
+#endif
+
+    var canvasSize = [
+      {{{ gpu.makeGetU32('config', C_STRUCTS.WGPUSurfaceConfiguration.width) }}},
+      {{{ gpu.makeGetU32('config', C_STRUCTS.WGPUSurfaceConfiguration.height) }}}
+    ];
+
+    if (canvasSize[0] !== 0) {
+      context["canvas"]["width"] = canvasSize[0];
+    }
+
+    if (canvasSize[1] !== 0) {
+      context["canvas"]["height"] = canvasSize[1];
+    }
+
+    var configuration = {
+      "device": WebGPU.mgrDevice.get(deviceId),
+      "format": WebGPU.TextureFormat[
+        {{{ gpu.makeGetU32('config', C_STRUCTS.WGPUSurfaceConfiguration.format) }}}],
+      "usage": {{{ gpu.makeGetU32('config', C_STRUCTS.WGPUSurfaceConfiguration.usage) }}},
+      // viewFormatCount
+      // viewFormats
+      "alphaMode": "opaque",
+      "width": canvasSize[0],
+      "height": canvasSize[1],
+      "presentMode": {{{ gpu.makeGetU32('config', C_STRUCTS.WGPUSurfaceConfiguration.presentMode) }}},
+    };
+    context.configure(configuration);
+  },
+
+  wgpuSurfaceGetCapabilities: (surfaceId, adapterId, capabilitiesPtr) => {
+    if (capabilitiesPtr !== 0) {
+      {{{ makeSetValue('capabilitiesPtr', C_STRUCTS.WGPUSurfaceCapabilities.formatCount, '1', 'i32') }}};
+      var format = navigator["gpu"]["getPreferredCanvasFormat"]();
+      var formatsPtr = WebGPU.Int_PreferredFormat[format]; // FIXME
```

Comment:
@kainino0x Do you know how I can properly fill `formats` in `wgpuSurfaceGetCapabilities()`?